### PR TITLE
Add stylelint for SCSS linting

### LINKED ISSUE #8618: Use ::before (pseudo elements) instead of :before (pseudo classes)
We should be using `::before` across the project instead of `:before`; browser support is there and it's more semantically correct. See: 
https://github.com/WordPress/gutenberg/pull/8588#discussion_r207855059

We should probably add a lint warning to the project as well so future devs don't use `:before` instead of `::before` by mistake, once this is closed.

### PR DESCRIPTION
**Pull request by noisysocks**
Supersedes https://github.com/WordPress/gutenberg/pull/1151. Closes #8618.

Adds the `npm run lint-css` and `npm run lint-css:fix` commands for SCSS linting, and configures `npm run lint` to automatically run `npm run lint-css`.

Linting is performed by [stylelint](https://github.com/stylelint/stylelint). I've configured stylelint to extend [stylelint-config-wordpress](https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress) which enforces the [WordPress CSS coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/).

To save time and reduce merge conflicts, I've disabled every rule that isn't totally trivial to fix (e.g. with a find and replace). We can work towards re-enabling these rules one by one over the medium term.

To test: 

1. `npm run lint-css`
2. Open the demo post and check for visual regressions
3. Check the diff and make sure `stylelint --fix` or one of my regular expressions didn't completely backfire

### CODE REVIEW

File: core-blocks/button/editor.scss [hunk 1]
```diff
@@ -1,9 +1,11 @@
 .editor-block-list__block[data-type="core/button"] {
+
```
Comment (noisysocks) [inline core-blocks/button/editor.scss]: Weird... why does it want spaces here, @ntwb?

File: packages/components/src/button/style.scss [hunk 2]
```diff
@@ -60,56 +61,62 @@
...
-				color( theme( primary ) shade( 30% ) ) 28%,
-				color( theme( primary ) shade( 30% ) ) 72%,
-				theme( primary ) 72%) !important;
-			border-color: color( theme( primary ) shade( 50% ) ) !important;
+				theme(primary) 28%,
+				color(theme(primary) shade(30%)) 28%,
+				color(theme(primary) shade(30%)) 72%,
+				theme(primary) 72%
+			) !important;
+			/* stylelint-enable */
```
Comment (ntwb) [inline packages/components/src/button/style.scss]: Note to self: Investigate supporting this in stylelint core

File: edit-post/components/meta-boxes/meta-boxes-area/style.scss [hunk 3]
```diff
@@ -28,7 +29,9 @@
 	/* Override Default meta box stylings */
 	#poststuff h3.hndle,
 	#poststuff .stuffbox > h3,
-	#poststuff h2.hndle { /* WordPress selectors yolo */
+	#poststuff h2.hndle {
+
+		/* WordPress selectors yolo */
```
Comment (ntwb) [inline edit-post/components/meta-boxes/meta-boxes-area/style.scss]: This is not ideal, it's an idiosyncratic stylelint issue that requires fixing 😞

**Review by ntwb — APPROVED** [review]
Awesome work @noisysocks 💯

**Comment by ntwb** [comment]
Reviewing the diff [without whitespace changes](https://github.com/WordPress/gutenberg/pull/8647/files?w=1) is helpful 👌

**Comment by gziolo** [comment]
Any idea why one of the unit tests failed on Travis?

File: core-blocks/button/editor.scss (hunk 1 above)
Comment (ntwb) [reply core-blocks/button/editor.scss]: A few reasons:

• Due to the number of permutations and combinations of _where_ and _where not_ new empty lines should occur, I've an issue to audit all of these at https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress/issues/65

• The lack off SCSS used currently in WordPress leading to not having an official "WordPress SCSS Coding Standards"

-----

Solution/Workaround:

Update the config used here to override the appropriate rules here in this PR and then follow up by updating `stylelint-config-wordpress`, either or both the CSS and SCSS configs and releasing a new version of `stylelint-config-wordpress`.

**Review by tofumatt — APPROVED** [review]
All of those weird new empty newlines aren't ideal, but I prefer automated, written rules I don't like over unwritten, unautomated ones I do like.

I'd be cool to fix all those newlines (eg https://github.com/WordPress/gutenberg/pull/8647/files#diff-44e5de5ab7a2465d8a970cbd777c427eR2). But hey, rules are cool!

**Comment by noisysocks** [comment]
Thanks all! I disabled `comment-empty-line-before`, `rule-empty-line-before`, and `at-rule-empty-line-before` as quick fix for the weird whitespace that was being added. Better to have the bare minimum than nothing at all 🙂

**Comment by noisysocks** [comment]
screen shot 2018-08-08 at 15 01 15

That ain't right... 😅

File: package.json [hunk 4]
```diff
@@ -161,11 +163,13 @@
 		"fixtures:server-registered": "docker-compose run -w /var/www/html/wp-content/plugins/gutenberg --rm wordpress ./bin/get-server-blocks.php > core-blocks/test/server-registered.json",
 		"fixtures:generate": "npm run fixtures:server-registered && cross-env GENERATE_MISSING_FIXTURES=y npm run test-unit",
 		"fixtures:regenerate": "npm run fixtures:clean && npm run fixtures:generate",
-		"lint": "npm run lint-js && npm run lint-pkg-json",
+		"lint": "npm run lint-js && npm run lint-pkg-json && npm run lint-css",
```
Comment (gziolo) [inline package.json]: Can we run it concurrently?

**Review by gziolo — APPROVED** [review]
Awesome one 👏 

I ❤️ `npm run lint-css:fix` the most!

File: package.json (hunk 4 above)
Comment (noisysocks) [reply package.json]: 👍 b97714574

**Comment by ntwb** [comment]
❤️

**Comment by aduth** [comment]
In case anyone is like me and was looking for editor integration for Sublime Text, here's the plugin for use with [SublimeLinter](https://github.com/SublimeLinter/SublimeLinter):

https://github.com/SublimeLinter/SublimeLinter-stylelint

**Comment by youknowriad** [comment]
Who uses sublime nowadays :trollface: ?

**Comment by tofumatt** [comment]
It's way faster than Atom! :trollface: :trollface: :trollface:

**Comment by youknowriad** [comment]
@tofumatt I fully deserve this one :P

**Pull request merged**